### PR TITLE
Implement Lazy sync on top of Eager sync

### LINKED ISSUE #12: Implement Lazy-sink strategy
use a key/value repository - Redis
recent content in memory
sync updates the things in memory
cache misses hit contentful directly
webhook causes the sync to get called.

### PR DESCRIPTION
**Pull request by gburgett**
continuation of #17 

This implements Lazy sync for the purpose of ID lookups, but punts on the problem of `find_by` lookups.  This gets us 90% of the way there for how we want to operate on Watermark Resources.

refs #18 - to continue with what we punted on

Fixes #12

### CODE REVIEW

File: lib/wcc/contentful/simple_client/typhoeus_adapter.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    TyphoeusAdapter::Response.new(
+      Typhoeus.get(
+        url,
+        params: query,
+        headers: headers
+      )
+    )
+  end
+
+  class Response
```
Comment (jpowell) [inline lib/wcc/contentful/simple_client/typhoeus_adapter.rb]: could be a `Struct`

File: lib/wcc/contentful/simple_client.rb [hunk 2]
```diff
@@ -35,6 +36,39 @@ def get(path, query = {})
...
+      http: ['http', '> 1.0', '< 3.0'],
+      typhoeus: ['typhoeus', '~> 1.0']
+    }.freeze
+
+    def self.load_adapter(adapter)
+      case adapter
+      when nil
+        ADAPTERS.each do |a, spec|
+          begin
+            gem(*spec)
```
Comment (jpowell) [inline lib/wcc/contentful/simple_client.rb]: I dig the auto discovery... whatever they have installed

File: lib/wcc/contentful/simple_client.rb [hunk 3]
```diff
@@ -58,12 +92,6 @@ def get_http(url, query, headers = {}, proxy = {})
     end
 
     def default_get_http(url, query, headers = {}, proxy = {})
-      if proxy[:host]
```
Comment (jpowell) [inline lib/wcc/contentful/simple_client.rb]: this a blank method for a reason?

File: lib/wcc/contentful/simple_client/typhoeus_adapter.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+gem 'typhoeus'
+require 'typhoeus'
+
+class TyphoeusAdapter
+  def call(url, query, headers = {}, proxy = {})
+    raise NotImplementedError, 'Proxying Not Yet Implemented' if proxy[:host]
+
+    TyphoeusAdapter::Response.new(
+      Typhoeus.get(
```
Comment (jpowell) [inline lib/wcc/contentful/simple_client/typhoeus_adapter.rb]: I guess we're always doing a `get`? Never a need to do a `delete` or `post`?

File: lib/wcc/contentful/store/lazy_cache_store.rb [hunk 5]
```diff
@@ -2,25 +2,75 @@
...
     end
 
     def find(key)
-      @cache.fetch(key) do
-        @store.find(key)
+      found =
+        @cache.fetch(key) do
+          # if it's not a contentful ID don't hit the API.
+          # Store a nil object if we can't find the object on the CDN.
+          (@cdn.find(key) || nil_obj(key)) if key =~ /^\w{22}$/
```
Comment (jpowell) [inline lib/wcc/contentful/store/lazy_cache_store.rb]: this always the format of a key?

File: lib/wcc/contentful/simple_client/typhoeus_adapter.rb (hunk 4 above)
Comment (gburgett) [reply lib/wcc/contentful/simple_client/typhoeus_adapter.rb]: Haven't had that use case yet

File: lib/wcc/contentful/store/lazy_cache_store.rb (hunk 5 above)
Comment (gburgett) [reply lib/wcc/contentful/store/lazy_cache_store.rb]: Yep all Contentful IDs are 22 character alphanumeric.  Though I definitely could understand if we don't want to build on that assumption...

File: lib/wcc/contentful.rb [hunk 6]
```diff
@@ -105,17 +105,24 @@ def self.init!
...
 
-    case configuration.content_delivery
-    when :eager_sync
-      store = configuration.sync_store
-
-      WCC::Contentful::Model.store = store
+    store =
+      case configuration.content_delivery
+      when :eager_sync
+        configuration.sync_store
```
[the remark this answers is not in the document]
Comment (gburgett) [reply lib/wcc/contentful.rb]: Your wish is my command

**Comment by gburgett** [comment]
This was merged into the wrong branch - we need to merge it into master.  Doing that in #23

**Pull request merged**